pkg/collector/corechecks/gpu/nvmlmetrics/collector.go

```diff
@@ -0,0 +1,161 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux
+
+// Package nvmlmetrics holds the logic to collect metrics from the NVIDIA Management Library (NVML).
+// The main entry point is the Collector struct, which is responsible for collecting metrics from the NVML library.
+// There are multiple subsystems that can be used to collect metrics from NVML: GPM, FieldInfo functions, and general
+// API functions. The collection of metrics is done by the subsystems, implementing the subsystemCollector interface.
+package nvmlmetrics
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// Collector is the main struct responsible for collecting metrics from the NVML library.
+type Collector struct {
```

Comment:
Even though they're similar, I wanted `subsystemCollector` to be an internal interface, as it deals with the device-specific collection (specially at first, as the `collect` function received a device) and it's not the same as the main `Collector` struct (e.g., the main collector doesn't have a name).

Reply:
but currently both "primary" methods have the exact same signature, and the only difference is the name method

Reply:
btw you can split to Collector interface and Namer interface if you wish the subsystem collectors and the main one to be separate. I find the name to be useful, because in the future you would have a similar one for AMD and Intel GPUs

Reply:
But do you also mean making the `subsystemCollector` interface public? I'm not sure about having a public interface for types that are only used internally in the package.

Reply:
I think it should be public, because the "compositeCollector" (in this example it is the current Collector struct) will be public and you will probably have an array (or map) of those in the global check.

the usage of that pattern actually clarifies the general design and structure of the code, because it is a common pattern in metrics collectors (e.g: prometheus is designed like that).